### ISSUE TITLE
Proposals: Lifecycle events as local events and broadcastLocal drop

### ISSUE THREAD

**Issue by pibi** [ice-services/moleculer#109]
I love the simplicity of the moleculer services and their structures.  I really appreciated the enhancements for events handling on 0.11.x . I think there's still some small room for improvements here, specifically about the local events handling, which have 2 broadcast methods: `broadcast('$eventName')` and `broadcastLocal('eventName')`. What is the actual use cases for these? What do you think to deprecate broadcastLocal altogether and just use `brodacast('$eventName')` for local events too?

Also, service lifecycle events are all local by definition. So, why there are those specialized methods here? What about three local events like `$service.created`, `$service.started` , `$service.stopped`'?

**Comment by icebob** [comment]
The `broadcastLocal` is important for internal events because it won't be transferred to remote nodes. Like `$services.changed`, `$node.connected`...etc. 
In a normal project recommended to use the `broadcast` and `emit`.

The lifecycle events is different as other, because it is only specified to a service instance, like events is not. It's also come from [Vue](https://vuejs.org/v2/guide/instance.html#Lifecycle-Diagram)

**Comment by pibi** [comment]
Yep, I see the lifecycle events coming from Vue, but what so you think if those events are also emitted locally to the same service? Here is the approach I'm using to fake this approach:

```
"use strict";

module.exports = {
	name: "test",
    events: {

	    "**" (payload, sender, event) {
            console.log(`TEST: Event '${event}' received from ${sender} node:`, payload);
        }
	},

 	started() {
        // emit a local event
        this.broker.emit("$service.started",null, [this.name]);

    },

    created() {
        // emit a local event
		this.broker.emit("$service.created",null, [this.name]);
    },

    stopped() {
        // emit a local event
		this.broker.emit("$service.stopped",null, [this.name]);
    }    	
}
```

**Comment by icebob** [comment]
You can do it in your services if you want. But the `emit` will transfer these events to other nodes too where maybe it is not relevant.

**Comment by pibi** [comment]
mmh, but from the doc: Every local events must start with $ (dollar sign). E.g.: $node.connected. If you call these events with emit or broadcast, they will send to only local services.

**Comment by icebob** [comment]
Hmm, it's true, sorry.

**Comment by pibi** [comment]
Same for the broadcastLocal proposal. If broadcast('$event') is always local, could it better to just use this and drop brodacastLocal?

**Comment by icebob** [comment]
Maybe, but in case of backward compatibility I won't remove it. But you can use broadcast instead of broadcastLocal. It does same thing if the event name starts with '$'.

**Comment by deividasjackus** [comment]
If I may put my two cents in... I'd vote for:
- Prefixing moleculer's internal events (e.g. `$node.connected`) with `$`
- Not prepending `$` to event names just because they're emitted/broadcast locally; using `broadcastLocal` / `emitLocal` instead.

My reasoning is the event name should not change based on consumer.

**Comment by pibi** [comment]
I think it is better to have a simple, clean and general purpose API which could be used as the foundation for higher level capabilities and I would rather  a clear and unique way to emit local vs remote events, at least at this development stage. Infact, @deividasjackus with '$' you are still changing the event handling logic based on consumer, as they are internal events only (local events within a node). 

By the way, I like the actual behavior, I just think it is better to get rid of duplications and have consistent API and events (i.e. the service lifecycle events proposal).

**Comment by deividasjackus** [comment]
@pibi I'm all for a simple, clean and general purpose API. Be that as it may, I am also for clean semantics.

An event name serves a clear purpose and should not have anything to do with routing logic (while it *can* specify the event name scope, e.g. `$node`).

**Comment by pibi** [comment]
@deividasjackus I agree with you about the clean semantics. Infact I think that calling actions locally are quite a corner-case in a microservice platform, isn't it?  I think they should be used for new moleculer modules or to develop middlewares than to high level usage.

Also, it would be interesting to know the reasons of the drop of the `emitLocal()` method, as I think it is quite similar to the broadcastLocal() scenario.

I know that other frameworks (mainly senecajs) uses pattern matching for routing, which is essentially a generalization of our "name.of.event" approach (it is a matter of serialization). My gut feeling is that the explicit `*Local()` methods don't help to avoid misunderstandings, unless we are not dropping the local `$` for non-internal local events. 

You could think this a bike shedding, but I had recently worked on one of the largest senecajs development, with more than 50 microservices, developed by a full remote team spanning 12 timezones. In this scenario we had quite a tough time with senecaJS and its declarative approach to microservices. Here I see moleculer more as the opinionated (as in configuration over code) framework to be used on large development scenarios (I really love the `broker.loadService('fileName')`, you can see the same approach on our senecajs reference architecture here: https://github.com/pibi/senecatest/blob/master/microservices/user/plugin.js#L17 )

**Comment by deividasjackus** [comment]
@pibi, 

- Calling local actions: I agree partially. If you subscribe to the concept of "microsystems" rather than "microservices" ([as proposed by Jonas Bonér](https://info.lightbend.com/ebook-reactive-microservices-the-evolution-of-microservices-at-scale-register.html)) local calls/events seem quite reasonable to me in certain situations, but it indeed is far from a common pattern (for now?).

- Explicit methods: I agree. `broadcast` and `emit` should route globally, `broadcastLocal` and `emitLocal` should route locally. Neither should not have event name related logic, expectations or restrictions. Disclosure: I'm still not as familiarized w/ v.10 changes as well as I'd like to be; was not aware `emitLocal` was dropped, let alone of the reason for it.

- Seneca: it's a shame you had to go through that, but great in the sense that you can now share and help Moleculer avoid the same mistakes 😃 Not bikeshedding at all IMO. A consistent core built on learnings from other projects' biggest annoyances isn't all that bad 😛

**Comment by icebob** [comment]
The `emitLocal` is not dropped, just renamed.
The `emit` logic in v0.11 is changed to use balanced event logic to support event-driven arch & event sourcing. The old emit logic was a simple broadcasting (to all services), so I renamed the old methods to `broadcast*` from `emit*`

The `emitLocal` in v0.10 is renamed to `broadcastLocal` in v0.11,
The `emit` in v0.10 is renamed to `broadcast` in v0.11.

I think the current logic is straightforward. All internal events got the `$` prefix. And the `$` prefix is reserved for internal core module events, like `$node.*` or `$services.changed`. So they don't conflict with the user custom events. 

The `emit` sends event with balancing & grouping which is essential for event drive & event sourcing. The event will be sent to only one instance per services.
The `broadcast` sends event to all services on all nodes (include local node & local services)
The `broadcastLocal` sends event to all local services only. It is used by core modules to emit the internal `$` events.
However in `emit` and `broadcast` have a protection, that they don't transfer events which is prefixed with `$`. 

In a nutshell:
In your own services, don't use the `$` prefix in event names when emit events. Use normal event names like `user.created`, `payment.charged`...etc and send them with `emit`. If you want to notify all services (e.g. `cache.clear`, or `config.changed`), use the `broadcast` method.
